components/core/src/fs.rs

```diff
@@ -773,7 +790,29 @@ impl AtomicWriter {
         debug!("Renaming {} to {}",
                self.tempfile.path().to_string_lossy(),
                &self.dest.to_string_lossy());
+
+        // `fs::rename` calls `MoveFileExW` on Windows, however the underlying implementation only
+        // utilizes the `MOVEFILE_REPLACE_EXISTING` flag which allows for file overwrite but no
+        // guarantee on durability. For this, we additionally pass `MOVEFILE_WRITE_THROUGH`.
+        // This causes the function to not return until the file is actually moved on the disk.
+        // Setting this value guarantees that a move performed as a copy and delete operation is
+        // flushed to disk before the function returns. The flush occurs at the end of the copy
+        // operation.
+        #[cfg(windows)]
+        unsafe {
+            if MoveFileExW(AtomicWriter::windows_u16s(self.tempfile.path()).as_ptr(),
+                           AtomicWriter::windows_u16s(&self.dest).as_ptr(),
+                           winapi::um::winbase::MOVEFILE_WRITE_THROUGH
+                           | winapi::um::winbase::MOVEFILE_REPLACE_EXISTING)
+               == 0
+            {
+                return Err(io::Error::last_os_error());
+            }
+        }
```

Comment:
This is great! I think a cross platform atomic rename could be useful in general. Would you mind extracting this code to a function? We could then export a function (maybe `atomic_rename`?) with the same signature for both windows and linux making it easy to use in the future.

Not for this PR but we at some point should probably check all the places we call `fs::rename`.

:thinking: I wonder if the rust team would accept a PR adding `MOVEFILE_WRITE_THROUGH` to their rename implementation?

Reply:
@davidMcneil I actually had this implementation then reverted the commit because I realized that a `atomic_rename` is effectively what we already have in the `AtomicWriter::finish` function, where I've added the platform specific bits.

https://github.com/habitat-sh/habitat/blob/edde4584add1d9dd5fc552f747c44bc95ffafb0f/components/core/src/fs.rs#L747-L781

It should generally be called via `atomic_write`:

https://github.com/habitat-sh/habitat/blob/edde4584add1d9dd5fc552f747c44bc95ffafb0f/components/core/src/fs.rs#L809-L814

Reply:
> I wonder if the rust team would accept a PR adding MOVEFILE_WRITE_THROUGH to their rename implementation?

👍 I'm definitely curious too. I'll open one up.

Reply:
> It should generally be called via atomic_write:

:thinking: I think that might be slightly different? To use that to rename a file I would write something like:

```rust
pub fn rename(src: &Path, dst: &Path) -> io::Result<()> {
    let contents = fs::read_to_string(src)?;
    atomic_write(dst, contents)?;
    fs::remove_file(src)
}
```

Between each of the three steps (reading, writing, and deleting) I think there are potentials for race conditions that an `atomic_rename` would guard against.

Reply:
@davidMcneil most definitely a brilliant idea!

I've added a commit for it: https://github.com/habitat-sh/habitat/pull/7783/commits/04145e4d68e1af2624af4fbb68f6c7c4c66daef1

Reply:
@jeremymv2 sorry, I think my last comment was confusing :sweat_smile: ? I was trying to say the example `rename` functionality was **NOT** atomic because there could be race conditions between each of those expressions. The true atomic rename I think should look something like:

```rust
// `fs::rename` calls `MoveFileExW` on Windows, however the underlying implementation only
// utilizes the `MOVEFILE_REPLACE_EXISTING` flag which allows for file overwrite but no
// guarantee on durability. For this, we additionally pass `MOVEFILE_WRITE_THROUGH`.
// This causes the function to not return until the file is actually moved on the disk.
// Setting this value guarantees that a move performed as a copy and delete operation is
// flushed to disk before the function returns. The flush occurs at the end of the copy
// operation.
#[cfg(windows)]
pub fn rename_windows<P: AsRef<Path>, Q: AsRef<Path>>(from: P, to: Q) -> io::Result<()> {
    // Helper function to transform a Path to a `LPCWSTR` Windows string data type. It is
    // essentially a null-terminated string of 16-bit Unicode characters.
    fn windows_u16s(s: &Path) -> Vec<u16> {
        s.as_os_str().encode_wide().chain(iter::once(0)).collect()
    }
    unsafe {
        if MoveFileExW(windows_u16s(from.as_ref()).as_ptr(),
                       windows_u16s(to.as_ref()).as_ptr(),
                       winapi::um::winbase::MOVEFILE_WRITE_THROUGH
                       | winapi::um::winbase::MOVEFILE_REPLACE_EXISTING)
           == 0
        {
            Err(io::Error::last_os_error());
        } else {
            Ok(())
        }
    }
}

/// atomic_rename is a cross platform  helper function for renaming a file atomically with
/// durability guarantees. It leverages the atomicity and durability features of atomic_write.
pub fn atomic_rename<P: AsRef<Path>, Q: AsRef<Path>>(from: P, to: Q) -> io::Result<()> {
    #[cfg(windows)]
    return fs::rename_windows(from, to);
    #[cfg(not(windows))]
    return fs::rename(from, to);
}
```

Reply:
It turned out to be a little tricky getting the trait bounds defined correctly because we also needed to call `AtomicWriter::sync_parent()` for `rename_unix()` so that the file changes were durable. However, I believe I've got it in https://github.com/habitat-sh/habitat/commit/406a125fb7fb87eb46de16fba18df77f40525a2e